tick/eval/node_value_acessors.go

```diff
@@ -0,0 +1,547 @@
+		// but the performance is neglibile - 8ns/op (this after trying to add NodeValueAccessor cache) for Benchmark_EvalBool_OneOperator_UnaryNode
+		nodeValueAccessor := createNodeValueAccessor(n.Node.Node)
+		if nodeValueAccessor == nil {
+			return false, fmt.Errorf("Got nil node value accessor for %v, this is invalid behavior - we expected the node to be BoolNode", n.Node.Node)
+		}
+
+		result, err := nodeValueAccessor.BoolValue(scope, executionAux)
+		if err != nil {
+			return false, err
+		}
+
+		return !result, nil
+	}
+
+	return false, fmt.Errorf("unary node: invalid operator %v", n.Node.Operator)
+}
+
+//-----------------------------------------
+// ReferenceNode NodeValueAcessor Implementation
+
+func (n *EvalReferenceNode) Value(scope *tick.Scope, executionAux ExecutionAux) (interface{}, error) {
+	value, err := scope.Get(n.Node.Reference)
+	if err != nil {
+		return nil, err
+	}
+
+	return value, nil
+}
+
+func (n *EvalReferenceNode) RegexValue(scope *tick.Scope, executionAux ExecutionAux) (*regexp.Regexp, error) {
+	refValue, err := n.Value(scope, executionAux)
+	if err != nil {
+		return nil, err
+	}
+
+	if regexValue, isRegex := refValue.(*regexp.Regexp); isRegex {
+		return regexValue, nil
+	}
+
+	return nil, ErrTypeGuardFailed
+}
+
+func (n *EvalReferenceNode) StringValue(scope *tick.Scope, executionAux ExecutionAux) (string, error) {
+	refValue, err := n.Value(scope, executionAux)
+	if err != nil {
+		return "", err
+	}
+
+	if stringValue, isString := refValue.(string); isString {
+		return stringValue, nil
+	}
+
+	return "", ErrTypeGuardFailed
+}
+
+func (n *EvalReferenceNode) Float64Value(scope *tick.Scope, executionAux ExecutionAux) (float64, error) {
+	refValue, err := n.Value(scope, executionAux)
+	if err != nil {
+		return float64(0), err
+	}
+
+	if float64Value, isFloat64 := refValue.(float64); isFloat64 {
+		return float64Value, nil
+	}
+
+	return float64(0), ErrTypeGuardFailed
+}
+
+func (n *EvalReferenceNode) Int64Value(scope *tick.Scope, executionAux ExecutionAux) (int64, error) {
+	refValue, err := n.Value(scope, executionAux)
+	if err != nil {
+		return int64(0), err
+	}
+
+	if int64Value, isInt64 := refValue.(int64); isInt64 {
+		return int64Value, nil
+	}
+
+	return int64(0), ErrTypeGuardFailed
+}
+
+func (n *EvalReferenceNode) BoolValue(scope *tick.Scope, executionAux ExecutionAux) (bool, error) {
+	refValue, err := n.Value(scope, executionAux)
+	if err != nil {
+		return false, err
+	}
+
+	if boolValue, isBool := refValue.(bool); isBool {
+		return boolValue, nil
+	}
+
+	return false, ErrTypeGuardFailed
+}
+
+//-----------------------------------------
+// FunctionNode NodeValueAcessor Implementation
+
+func (n *EvalFunctionNode) Value(scope *tick.Scope, executionAux ExecutionAux) (interface{}, error) {
+	args := make([]interface{}, len(n.Node.Args))
+	for i, argNode := range n.Node.Args {
+		argNva := createNodeValueAccessor(argNode)
```

Comment:
Actually after thinking about this - we handle this as well, "value" + "another" will be a BinaryNode which we support.